static/app/views/settings/account/notifications/notificationSettings.tsx

```diff
@@ -0,0 +1,292 @@
+import {NOTIFICATION_SETTING_FIELDS} from 'app/views/settings/account/notifications/fields2';
+import {
+  getChoiceString,
+  getFallBackValue,
+  groupByOrganization,
+  isGroupedByProject,
+  providerListToString,
+} from 'app/views/settings/account/notifications/utils';
+import Form from 'app/views/settings/components/forms/form';
+import JsonForm from 'app/views/settings/components/forms/jsonForm';
+import {FieldObject} from 'app/views/settings/components/forms/type';
+import SettingsPageHeader from 'app/views/settings/components/settingsPageHeader';
+import TextBlock from 'app/views/settings/components/text/textBlock';
+
+type Props = {
+  notificationType: string;
+  organizations: Organization[];
+} & AsyncComponent['props'];
+
+type State = {
+  notificationSettings: {
+    [key: string]: {[key: string]: {[key: string]: {[key: string]: string}}};
+  };
+  projects: Project[];
+} & AsyncComponent['state'];
+
+class NotificationSettings extends AsyncComponent<Props, State> {
+  getDefaultState(): State {
+    return {
+      ...super.getDefaultState(),
+      notificationSettings: {},
+      projects: [],
+    };
+  }
+
+  getEndpoints(): ReturnType<AsyncComponent['getEndpoints']> {
+    const {notificationType} = this.props;
+
+    const query = {type: notificationType};
+    const endpoints: ReturnType<AsyncComponent['getEndpoints']> = [
+      ['notificationSettings', `/users/me/notification-settings/`, {query}],
+    ];
+    if (this.isGroupedByProject()) {
+      endpoints.push(['projects', '/projects/']);
+    }
+    return endpoints;
+  }
+
+  isGroupedByProject() {
+    /** We can infer the parent type by the `notificationType` key. */
+    const {notificationType} = this.props;
+    return isGroupedByProject(notificationType);
+  }
+
+  getParents(): Organization[] | Project[] {
+    /** Use the `notificationType` key to decide which parent objects to use */
+    const {organizations} = this.props;
+    const {projects} = this.state;
+
+    return this.isGroupedByProject() ? projects : organizations;
+  }
+
+  getUserDefaultValues = (): {[key: string]: string} => {
+    /**
+     * Get the mapping of providers to values that describe a user's parent-
+     * independent notification preferences. The data from the API uses the user
+     * ID rather than "me" so we assume the first ID is the user's.
+     */
+    const {notificationType} = this.props;
+    const {notificationSettings} = this.state;
+
+    return (
+      Object.values(notificationSettings[notificationType]?.user || {}).pop() || {
+        email: getFallBackValue(notificationType),
+      }
+    );
+  };
+
+  getParentData = (): {[key: string]: string} => {
+    return Object.fromEntries(
+      this.getParents().map(parent => [
+        parent.id,
+        Object.values(this.getParentValues(parent.id))[0],
+      ])
+    );
+  };
+
+  getStateToPutForProvider = changedData => {
```

Comment:
hmm, I feel like this might be a more complicated approach to update the state on change. If the API returns the exact shape, you can just call the `this.fetchData` method, to get the latest and re-render

Reply:
This function calculates the diff of the old state and new state and sends that diff to the API. Since doing so gives us the new state, I'm just using that new state to update the UI.